src/module.rs

```diff
@@ -136,6 +135,7 @@ enum_rename!{
 #[derive(Debug, PartialEq, Eq)]
 pub struct Module<'ctx> {
     data_layout: RefCell<Option<DataLayout>>,
+    dibuilder: RefCell<Option<DebugInfoBuilder<'ctx>>>,
```

Comment:
It's not clear to me, why does the DIB need to live in the Module?

Reply:
To call `dispose` when dropped. I was thinking of tracking if `finalize` has been called inside DIB in order to decide if it needs to be called before `dispose`.

I'd also like to call `finalize` automatically whenever code generation happens but that doesn't seem possible, it should be called even whenever a verification pass is performed (for example in function or module verify) and there's no clear way to do that.

Reply:
> To call dispose when dropped.

Why can't we call dispose on `DebugInfoBuilder`'s drop impl, though?

>  I was thinking of tracking if finalize has been called inside DIB in order to decide if it needs to be called before dispose.

This is a valid option. Especially if LLVM doesn't have a getter for this already...

> I'd also like to call finalize automatically whenever code generation happens but that doesn't seem possible, it should be called even whenever a verification pass is performed (for example in function or module verify) and there's no clear way to do that.

Should be, or must be? If nothing bad happens (ie UB) if you neglect to do so then I think the API user should be responsible for calling finalize prior to codegen / passes / etc.

Reply:
Also, does calling finalize make the DIBuilder immutable (ie can't make changes to it?). If we we might want two different states. Like `DIBuilder<Unfinalized>` and `DIBuilder<Finalized>` so that a finalized DIB can't call its methods (I'm assuming bad things happen in that scenario here...)

Reply:
> Why can't we call dispose on DebugInfoBuilder's drop impl, though?

I made it `Copy` when I was testing stuff and `Copy` implementors can't `Drop`. I can change that back.

 > Should be, or must be? If nothing bad happens (ie UB) if you neglect to do so then I think the API user should be responsible for calling finalize prior to codegen / passes / etc.

Definitely *must*, otherwise debug info is in incomplete state and invalid. 

> Also, does calling finalize make the DIBuilder immutable (ie can't make changes to it?). If we we might want two different states. Like DIBuilder<Unfinalized> and DIBuilder<Finalized> so that a finalized DIB can't call its methods (I'm assuming bad things happen in that scenario here...)

Haven't tested that yet. If it is prohibited to use DIBuilder after finalize, the function can take `self` instead of a reference (if it isn't `Copy`) and consume the builder.

Reply:
> I made it Copy when I was testing stuff and Copy implementors can't Drop. I can change that back.

I see, I didn't notice that it was marked Copy.

> Definitely must, otherwise debug info is in incomplete state and invalid.

Is this only an IR issue though? We can't prevent all runtime IR errors. (We should certainly try to prevent as much as possible, but not everything can be prevented). I'm more concerned about memory safety here. Writing some tests and/or looking through LLVM source would probably be helpful (if you get stuck, I can try and take a look too)

> Haven't tested that yet. If it is prohibited to use DIBuilder after finalize, the function can take self instead of a reference (if it isn't Copy) and consume the builder.

Yeah that'd be a better option iff there's nothing to be called on a finalized builder

Reply:
The fact that finalize and dispose are two separate functions makes me suspicious that there's something you can do with a finalized builder.